### DISCUSSION TITLE
Request to add raw markdown writing view

### DISCUSSION THREAD

**Discussion by wohenbushuang** [outline/outline#3326] [Ideas]
Request new feature to switch the editor view modes, such as raw markdown, splited double columns for synchronized scroll preview, and WYSIWYW as now. Similar products: [Hedgedoc](https://demo.hedgedoc.org/features?both), Github markdown editor, etc.

## Reason
For now, it's quite hard to do some writing jobs on the rendered texts. For example:
* append to link text
* prepend to bold text

Providing a pure markdown code editor somewhere will enhance the ability for delicate operation.

Furthermore, it will be easier to add [Latex](https://demo.hedgedoc.org/features#MathJax) and [diagrams](https://demo.hedgedoc.org/features#Diagrams) support in the future.

**Comment by tommoor** [comment]
Hi @wohenbushuang – honestly we won't be supporting a raw markdown editor in this project, we're doubling down on making the editor collaborative and introducing features like mentions, inline comments, rich embeds and other things that just aren't compatible with the limited markdown spec.

I think your frustrations around editing the edges of marks could be addressed without this change however.

**Reply by senpro-ingwersenk** [reply]
In this case: How much "Markdown" is outline then, if you can not edit in raw Markdown?

I started work at a new company and we run Outline as a documentation basis. Thing is, there are weird behaviours with headings and alike, which would be very easily verifiable if I could view the source. And the source would technically be raw markdown.

And this is what confuses me; Outline is said to be a markdown wiki system, but does not show raw markdown to fix the generated markdown?

Thanks for your hard work! ^^

**Comment by wohenbushuang** [comment]
@tommoor First of all, I've updated to v0.65.1. Outline is now my daily notebook application, and I can easily share certain page to friends, or a scoped structal document to dev coworkers! 😄 

However, until now I can still find the editing experience weird sometimes. Below is one example I met today. Besides, in daily usage, there exists other bad experience cases. These were hard to screen recorded, video clipped and annotated for new issues due to the time cost, while also hard to describe the reproduction procedures and phenomenons in words.

![1](https://user-images.githubusercontent.com/15613184/179784250-d361db14-c436-4507-bd52-47bbceb0d6cb.gif)

I have to say that raw markdown editor is user friendly from some aspects, since the styling is also being controled within raw code textarea, as well as editing and rendering are saperated. All editing actions could be done on the raw code, regardless the rendered result of the previous step, and won't be interruptted by block rearrangements.

But for a WYSIWYW, these days I realized that the troubles I suffered are result from outline hides the menu toolbar. The styling functions are now only available with markdown syntax, but it will be rendered and disappeared once I completed; or by the hovering toolbar, which has limited options. ProseMirror and other WYSIWYW editors will have a menu toolbar or context menu support containing all styling function options, and user can easily find their need/see styling status, then click them (eg., cancel enumeration in the above example) whenever they want. I think that the hovering toolbar is also useful and convenient (maybe inspired from MS Word 2013+ ?), but it is for a shortcut. You cannot put every styling functions into it due to the limited room (BTW, the editing experience on mobile screen is awful). So, can you imagine the difficulty to use MS Word without menu toolbar and context menu?

![image](https://user-images.githubusercontent.com/15613184/179911404-0dcbddf8-d0b3-4bcd-b996-a872ef13c027.png)

![image](https://user-images.githubusercontent.com/15613184/179912224-574b6314-9b02-4615-a5a5-ebc90fcb91ef.png)

![image](https://user-images.githubusercontent.com/15613184/179912268-e3fac0fb-4f9c-4697-81b2-803631b62ec4.png)

In a word, I recommend to add the menu toolbar back, maybe globally fixed top or bottom would be nice.

**Comment by socketpair** [comment]
We have two categories of people:
1. programmers - it's much faster for them to write markdown text instead of WYSIWYG
2. Managers - it's much faster for them to use WYSIWYG
3. Technical writers - it's much easier for them to cleanup text after managers

PLEASE add raw markdown editing.

**Reply by threema-lenny** [reply]
Being part of the _programmers_ category, I have no issues stripping weird additions to Markdown due to proprietary features if we can get source editing. I also have no issues with this being an experimental function if we can get source editing. But, my god, is WYSIWYG driving me nuts.

I must say, this product is advertising with "Markdown compatibility" and no source editing is, at the very least, misleading to me.

**Reply by tommoor** [reply]
> I have no issues stripping weird additions to Markdown due to proprietary features if we can get source editing

I'm not sure what you mean here. The problem is that markdown cannot represent much of what a modern collaborative editor needs to achieve – embeds, resizable table columns, comments, multiplayer editing, the list goes on… this is why I don't think it's possible for the platform to support both 🤔 

But hey, if someone has a fantastic solution I'm open to ideas

**Reply by threema-lenny** [reply]
It probably depends on the underlying storage format you've chosen that achieves these features which I haven't looked at so far.

If this is an extended markdown with, for example, some proprietary inline HTML-ish extensions that are parsed by the WYSIWYG engine to achieve things such as comments, sized tables, etc., then I wouldn't mind editing that source because I could strip those extensions or leave them as-is. And I could totally live without collaborative editing in source editing mode. Sure, it does sound like a hack (because it kinda is) but I'm truly that desperate if the source is still editable within reason.

If it isn't some kind of extended markdown underneath for storage, then it's probably not even possible.

**Reply by wohenbushuang** [reply]
We programmers need a panel to edit the source to deal with some (many, actually) strange bugs before the editor is strong enough. It can be like the dev tool in Chrome, hiden by default and used by professional users.

**Reply by socketpair** [reply]
Look at GitHub itself. It allows to edit Markdown directly. Sometimes WYSIWYG editor does strange things, and it's not always easy to fix them. So, please, again, add text area with raw markdown.

**Comment by socketpair** [comment]
@tommoor Please take a look at the issue. It's a very useful thing for us.

**Comment by matthijskooijman** [comment]
One viewpoint that I have not seen so far is about copy-pasting markdown from elsewhere. When I copy markdown source for e.g. a bullet list into outline, it ignores the - marks and just inserts them literally. To paste something into outline, I need a rendered (HTML) version and copy-paste that, which seems to work but is probably also more likely to include extra formatting garbage (that might be stripped if outline cannot represent it, but might also linger if outline can, I guess).

**Reply by tommoor** [reply]
It depends on where you copy from, if there is HTML on the clipboard that will be used which can result in the behavior you're seeing. To ensure it's detected as Markdown use "Paste and match style", or the Windows equivalent

<img width="236" alt="image" src="https://github.com/user-attachments/assets/a83256bf-f4c6-4026-97b5-e016c10ce02d">

**Reply by matthijskooijman** [reply]
In this particular case I'm pasting from Joplin or Hedgedoc. I do not have that "Paste and Match Style" option here (I'm assuming it should be in the browser context menu?) in Firefox on Linux, just Paste and Paste without formatting.

But the point of my comment was that if you could edit raw markdown, then you can reliably paste markdown from any source, without any issues that might (not) occur "depending on where you copy from". For this particular usecase (and I think some others as well), I even think it would be acceptable if such a raw markdown editing mode would *not* be collaborative, and maybe even only works for a subset of documents (or would lose some formatting). I also imagine that this is easier said than done :-)

**Reply by tommoor** [reply]
> I do not have that "Paste and Match Style" option here (I'm assuming it should be in the browser context menu?) in Firefox on Linux, just Paste and Paste without formatting.

"Paste without formatting" would be the Linux equivalent which will detect markdown and convert to the editor format.

**Reply by matthijskooijman** [reply]
If so, that does not really help me. If I copy markdown source for a bullet list and use "Paste without formatting" that ends up with literal dashes in the resulting text, not a bullet list. If I copy the rendered markdown (from Joplin) bulletlist, and use "Paste without formatting"  I get only the text, no dashes, no bullet list in the result (as I would expect, so "Paste without formatting" does not really help).

**Reply by tommoor** [reply]
Strange, perhaps something different with the Linux implementation of that then

**Reply by tommoor** [reply]
One thing that might help to debug this is to install some sort of clipboard viewer and see exactly what is contained on the clipboard for pasting.

**Comment by chulucninh09** [comment]
I surely need markdown. We are integrating outline with our AI workflow and it should support markdown as people copy paste markdown around a lot. I understand outline also provide AI features but we are using our own AI workflow due to the complexity in our workflow. So markdown should be very good feature. Right now we have to import markdown to outline and it really makes us slow down a lot

**Comment by DaxtonD** [comment]
One of the things I really appreciate about Outline is its rich-text editing experience. While it uses Markdown under the hood, it doesn’t feel like a Markdown editor — and that’s a big part of why I chose it over other tools.
To me, Markdown can feel a bit outdated and unintuitive, especially for users who prefer a visual or WYSIWYG-style workflow. I totally understand that some users love Markdown, and that’s valid — but I’d strongly prefer if it remained optional rather than becoming a core part of the editing interface.
If Markdown is your primary workflow, there are already a number of excellent tools built specifically for that, like Obsidian, Zettlr, Logseq, Typora, and Joplin. 
What makes Outline stand out, in my view, is its clean, collaborative, and user-friendly design especially for teams that aren’t necessarily Markdown-focused. I really hope that continues to be a core part of Outline.

**Reply by threema-lenny** [reply]
I believe the vast majority here would be absolutely fine with raw editing being optional. But please make it exist!

The fiddlyness of tables, the WYSIWYG editing just not doing what I want it to, copy & paste being sometimes behaving bonkers is overall just so cumbersome, it puts me off every day.

This isn't meant disrespectful towards the developer either: As a person working on a messaging solution, I know how horrible and underspecified HTML `contenteditable` behaves. But exactly for that reason, the option for raw editing is dearly missed.

**Reply by wohenbushuang** [reply]
> If Markdown is your primary workflow, there are already a number of excellent tools built specifically for that, like Obsidian, Zettlr, Logseq, Typora, and Joplin.

If WYSIWYG is your primary workflow, there are already a number of excellent tools built specifically for that, like MS office 365, Google doc and numerous other rich text editors without collaborative editing.

Also, please read through the post carefully before your reply. The request to **add** raw markdown view has been stated above, as well as the shit experience with WYSIWYG.

**Reply by DaxtonD** [reply]
Just to clarify — I did read the post carefully. My response was not an attempt to shut down discussion but to share a different user perspective, particularly from someone who chose Outline because of its current editing model. There’s room for multiple workflows in this conversation, and it’s valuable for maintainers to hear a broad range of experiences.

Your tone in response felt a bit dismissive, which is unfortunate. I think we can all agree that productive discussions are best when respectful, even if we disagree.

To your point: yes, WYSIWYG workflows are supported by tools like Google Docs and Office 365, but these don’t offer the same features or structured documentation environment that Outline does, nor are they open-source or self-hostable. That’s what makes Outline unique — it blends a modern UX with structured knowledge management, and that’s why I use it.

If Markdown support is added as an optional raw view or developer tool, that could be a great compromise for technical users who want full control. But for folks like me, who prefer a distraction-free visual editor, the current model is part of what makes Outline a good fit.

Let's keep this conversation constructive — the goal, after all, is to improve a tool we all clearly care about.

**Comment by atomlab** [comment]
Outline is an excellent tool for collaboration. Many thanks to the developers for creating it. In my case, employees who prefer WYSIWYG editors write the documentation, and then we automatically export it via the API to Markdown and generate a static site (for example, using Vitepress). I need to be able to add object properties through Markdown directly in Outline, for example:

```yaml
---
order: 1
layout: doc
tags: guides
---
```
If Outline supported editing raw Markdown, it would significantly expand its range of use cases and make automation much easier for engineers.

**Comment by strangerinusall** [comment]
I think having a separate 'raw' mode where markdown would be shown instead of rendered text would be very much appreciated. The beauty of Markdown is that it's just pure text and can be easily parsed by computers. Not being able to edit or at least see the raw Markdown easily makes it harder to work with it. 

Also, such mode does not have to interfere with those people who only want to see the rendered mode. It's just a matter of having a hidden button and a keyboard hotkey combination to enable/disable it.

**Comment by saxtron3000** [comment]
I'd like to see a solution to the H4, H5 and H6 issues I am having. Since they are not in the tool bar, they are tricky to access and edit. I can't type in the formatting but a copy and past from another .md works. Super confusing and would love to switch to the raw .md to clear all this up.

**Reply by tommoor** [reply]
They're not there because we do not recommend using headings to this level of nesting, everything beyond H4 is displayed the same.

**Reply by saxtron3000** [reply]
Thanks for the reponse. 

I can get why you would not recommend it, but why disable it? 

Right now my work around is to copy paste `##### Test-Heading` from a .md file and H5 works. Typing it directly in doesn't work. I am using a PDF generator that requires H5 and H6 (it's complicated) but with these disabled, workflow is tricky. 

Any suggestions?
